rfcs/0086-nixos-generation-apis.md

```diff
@@ -0,0 +1,195 @@
+---
+feature: nixos-generation-apis
+start-date: 2021-02-09
+author: Michael Lohmann <mial.lohmann@gmail.com>
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: https://github.com/NixOS/nixpkgs/pull/105910, https://github.com/NixOS/nixpkgs/issues/24374
+---
+
+# Summary
+[summary]: #summary
+
+On a typical system there are lots of generations. However rolling back to a
+specific one without rebooting is hard and it is even more difficult to get an
+overview of the available generations. This RFC proposes adding
+`nixos-rebuild --generation <number>` for rolling back to a specific generation
```

Comment:
```suggestion
`nixos-rebuild switch --generation <number>` for rolling back to a specific generation
```

Reply:
I thought about this and my thoughts were, to keep the interface as similar to the `--rollback` as possible. Also you could just as well `test`/ a generation. I thought about making `switch` the default (overridable) action but my main doubts came when @edolstra said
> There is an issue somewhere about replacing `nixos-rebuild` with a `nixos` command

and then I think `nixos --generation <n>` would be missing a verb ^^ (but maybe that is just personal preference...)

In general I got a bit unsatisfied with the name `switch` when I realised that the default behaviour is really doing a `build-and-switch` and here we really just want to switch to something already existing. This drove me a bit crazy when I learned about profiles while implementing this and I didn't know how to distinguish "switching to an existing profile" vs "creating a new generation of a profile". Actually I am thinking now, that the `--generation` flag is the only current possibility to just switch profiles - but I am probably overlooking something...

Reply:
My bad. I forgot that the options are defined on the top-level.

In some sense, `--generation <n>` is a more general version of `--rollback` so I guess both flags should be mutually-exclusive.

Reply:
I didn't enforce it with error messages, but the `--generation` is in an `else`-clause that doesn't get executed if both are given. Should I throw errors on all incompatible / useless combinations of commands? Right now they just don't do anything (i.e. `nixos-rebuild list-generations --rollback` will just list them and ignore the rest).